asm-utils/src/data_storage.rs

```diff
@@ -2,76 +2,157 @@
 
 use std::collections::BTreeMap;
 
-use crate::{data_parser::DataValue, utils::next_multiple_of_four};
+use crate::{
+    data_parser::DataValue,
+    utils::{alignment_size, next_aligned},
+};
 
 pub enum SingleDataValue<'a> {
     Value(u32),
-    LabelReference(&'a String),
-    Offset(&'a String, &'a String),
+    LabelReference(&'a str),
+    Offset(&'a str, &'a str),
 }
 
-pub fn store_data_objects<'a>(
-    objects: impl IntoIterator<Item = &'a (String, Vec<DataValue>)> + Copy,
+struct WordWriter<'a, 'b> {
+    data_writer: &'a mut dyn FnMut(u32, SingleDataValue) -> Vec<String>,
+    partial: u32,
+    current_pos: u32,
+    generated_code: Vec<String>,
+
+    latest_label: Option<&'b str>,
+}
+
+impl<'a, 'b> WordWriter<'a, 'b> {
+    fn new(
+        starting_pos: u32,
+        data_writer: &'a mut dyn FnMut(u32, SingleDataValue) -> Vec<String>,
+    ) -> Self {
+        // sanitary alignment to 8 bytes
+        let current_pos = next_aligned(starting_pos as usize, 8) as u32;
```

Comment:
where's this 8 coming from?

Reply:
@chriseth do you know why this is aligning to 8 bytes?

Reply:
I think this is future proof. The calling code pass a hardcoded `starting_pos` value that is already aligned to 8 bytes.